File: Archived/nlHydro/Functions/waveElevation.py
"""Wave elevation functions."""

import numpy as np


def main(elevationType, nF, nDir, waveDirections, time, centres, ramp, phase, omega, k, zeta):
	

    if elevationType == "mean":
        eta = np.zeros(len(centres[0, :]))
    elif elevationType == "instantaneous":
        eta = np.zeros(len(centres[0, :]))
        for i in range(nDir):
            for j in range(nF):
                X = k[j] * (centres[0, :] * np.cos(waveDirections[i]) + centres[1, :] * np.sin(waveDirections[i]))
                eta += ramp * zeta[i, j] * np.cos(omega[j] * time - X + phase[i, j])
                
    return {"status": "ok", "eta": eta}

if __name__ == "__main__":
	main()

File: applications/Validation/NL_hydro/nlHydro_c/wave_elevation.c
/**
 * wave_elevation.c
 * Free-surface wave elevation calculation.
 *
 * Mirrors waveElevation.py:
 *   eta[i] = sum_{dir,freq} ramp * zeta[dir,freq]
 *            * cos( omega[freq]*t - k[freq]*(x*cos(dir)+y*sin(dir)) + phase[dir,freq] )
 *
 * Array layouts (column-major, matching transposed numpy arrays):
 *   centres[3 x N]:       centres[row + 3*col]
 *   phase  [nDir x nF]:   phase [dir + nDir*freq]
 *   zeta   [nDir x nF]:   zeta  [dir + nDir*freq]
 */

#include <math.h>
#include <string.h>

#include "wave_elevation.h"

void wave_elevation(
    ElevationType  type,
    int            nF,
    int            nDir,
    const double   waveDirections[],
    double         time,
    double   centres[][3],
    int            N,
    double         ramp,
    double   phase[][nF],
    double   omega[],
    double   k[],
    double   zeta[][nF],
    double         eta_out[])
{
    /* Initialise output to zero */
    memset(eta_out, 0, N * sizeof(double));

    if (type == ELEVATION_MEAN) {
        /* eta remains zero everywhere */
        return;
    }

    /* ELEVATION_INSTANTANEOUS */
    for (int i = 0; i < nDir; i++) {
        double cos_dir = cos(waveDirections[i]);
        double sin_dir = sin(waveDirections[i]);

        for (int j = 0; j < nF; j++) {
            double amp = ramp * zeta[i][j];
            double ph  = phase[i][j];
            double kj  = k[j];
            double wj  = omega[j];

            for (int p = 0; p < N; p++) {
                double x = centres[p][0];
                double y = centres[p][1];
                double X = kj * (x * cos_dir + y * sin_dir);
                eta_out[p] += amp * cos(wj * time - X + ph);
            }
        }
    }
}
